### ISSUE TITLE
Can display_current_settings.c be moved to the "shared" directory?

### ISSUE THREAD

**Issue by tbohn** [UW-Hydro/VIC#354]
There doesn't appear to be anything in display_current_settings.c that is not in vic_driver_shared.h or vic_def.h.  Which means that we could theoretically just have one version of display_... in the drivers/shared/src directory.

Are the driver-specific versions of display_current_settings.c intended to eventually diverge?

At the moment, their presence just makes code development more difficult as updating for addition/removal of options requires updating each file exactly the same way. Already the files are slightly different in terms of format strings, which variables are displayed (admittedly global.domain doesn't really have an applicable value in the classic driver, but this could be handled with logic around the print statement) and I suspect most of these differences were unintended.

**Comment by jhamman** [comment]
I'm not sure how I feel about this.  What we're trying to ovoid is having a structure that looks like this anywhere in `vic_run` or `drivers/shared`:

```C
if strcmp(VIC_DRIVER, "Image") {
    // Do something
}
else if strcmp(VIC_DRIVER, "Classic") {
    // Do something else
}
...
```

I think the reason each driver has its own `display_current_settings.c` is that we only want to display settings relevant to the driver being used. There are settings in each driver that are not useful to the others.  We could change this, but I think we'd need a good reason. @bartnijssen and @wietsefranssen any thoughts?

**Comment by tbohn** [comment]
I guess I don't understand the advantage of having several copies of mostly redundant files, or the disadvantage of the "if" logic.  The have already diverged and this type of thing makes it more difficult to ensure that one has updated all the files exactly the same way.

**Comment by jhamman** [comment]
I think the right way to do this is to refactor the `display_current_settings.c` of all drivers into smaller functions, the parts that need to be different can be defined by the driver, everything else can combined and put into `driver/shared`.

**Comment by tbohn** [comment]
You can assign this to me if you'd like - I'll do it the way you suggest.

**Comment by jhamman** [comment]
Okay, so how about this plan.  
- Each driver keeps its own `display_current_settings.c`
- `driver/shared` gets a `print_current_settings.c` where a series of functions will consolidate code from the `classic/image/python` drivers that prints the settings that are shared among the drivers.  Everything else stays in the driver specific function.

**Comment by tbohn** [comment]
Sounds good to me.  So, can you clarify what you mean by "series of functions" - do you want to create several sub-functions, perhaps for printing different sections/categories of settings?  Or do you mean something else?

**Comment by jhamman** [comment]
> Sounds good to me. So, can you clarify what you mean by "series of functions" - do you want to create several sub-functions, perhaps for printing different sections/categories of settings?

That's exactly what I mean.

**Comment by tbohn** [comment]
OK, will do.

**Comment by wietsefranssen** [comment]
I agree with Joe's suggestion. So keeping `display_current_settings.c` in each driver and a series of small functions. Although this `display_current_settings.c` issue seems not too important, I think is this a very important -philosophical- discussion because we will have to deal with it in others parts of the code in the future.
In general I would suggest: make function as specific and small as possible.

**Comment by tbohn** [comment]
So, if I understand correctly, the approach we want to take in general is to implement conditional logic that depends on the driver *not* as "if" statements that appear in the code, but as alternate versions of the files - essentially using "make" to do the conditional logic once at compile time.

Now that I understand the philosophy, I'm fine with it.  Just out of curiosity, is the reason for this approach the avoidance of having the code check its own version at run time?  Is it a performance hit?

**Comment by bartnijssen** [comment]
Same here - I prefer the implementations suggested by Joe and Wietse. We don't want if statements to distinguish the different drivers. For me, the main reason is not performance-based (the hit of additional `if`s for version checking is not meaningful compared to the remaining computations). For me the reason is that in reality the drivers will likely be maintained by different communities, so there is no reason to force them to stick to the same conventions in the drivers. People who exclusively use the `classic` driver are not likely to test that changes they make in the driver don't break things for other modes. The only parts that have to be used in all the versions are contained within the `vic_run
` functionality.
